allennlp/data/transforms.py

```diff
@@ -0,0 +1,531 @@
+from typing import Dict, Tuple, List, Iterable, Generic, TypeVar, Deque, Union
+import itertools
+from collections import deque, defaultdict
+import random
+
+import torch
+from torch.utils.data import IterableDataset as IterableTorchDataset
+
+
+from allennlp.common.registrable import Registrable
+from allennlp.data.instance import Instance
+from allennlp.data.vocabulary import Vocabulary
+
+from allennlp.data.fields import MetadataField
+
+from allennlp.common.util import lazy_groups_of
+
+
+Batched = Iterable[Instance]
+InstanceOrBatch = TypeVar("InstanceOrBatch", Iterable[Instance], Instance)
+
+
+class DatasetFromGenerator(IterableTorchDataset):
+    def __init__(self, generator: Iterable[Union[Iterable[Instance], Instance]]):
+        self.generator = generator
+
+    def __iter__(self) -> Iterable[Union[Iterable[Instance], Instance]]:
+
+        for x in self.generator:
+            yield x
+
+
+class Transform(IterableTorchDataset, Generic[InstanceOrBatch], Registrable):
+    def transform(self, dataset: Iterable[Instance]) -> Iterable[InstanceOrBatch]:
+        """
+        Describes a transformation from either:
+
+        Instance -> Instance (e.g inplace mutation, indexing)
+        Instance -> Iterable[Instance] (batching, reading X number of instances into memory)
+        """
+
+        raise NotImplementedError
+
+    def transform_batch(self, batches: Iterable[Batched]) -> Iterable[Batched]:
+
+        # TODO Maybe switch this so that the base class just yields and the
+        # subclasses which are generic over Batched yield from
+        for batch in batches:
+            yield from self.transform(batch)  # type: ignore
+
+    def __call__(self, dataset: Iterable[Instance]) -> Iterable[InstanceOrBatch]:
```

Comment:
I understand why this functionality is nice for writers of transforms, but I can't shake the feeling that we have an opportunity here to make the `Transform` abstraction a bit cleaner. Essentially, the top-level `Transform` could be completely generic. It maps A -> B. One use case could be that a transform could take in URL strings and output iterables of instances. Of course, for the bulk of the transforms this dispatching magic is still useful, so we would have an `InstanceTransform` (or something) superclass that preserved it.

Reply:
I added this - I think it will be good for people who know what they are doing (i.e they know that the types of the transforms that get chained together have to match up exactly). The main reason for this dispatching logic is so that any of the `Transforms` can be chained together in arbitrary order and not break. But it would be great if we could just have a dataset of url strings, and then have transforms which convert them into instances. It doesn't exactly play nicely with the rest of the library (e.g we use `text_to_instance` from the dataset reader in other places, such as the predictor, and it's not clear how that would work if the thing that produces instances is now a Transform) but definitely it's cleaner.

Reply:
Honestly, I'm okay with leaving this out for now too. My initial worry was that if we don't do it now, we'll paint ourselves into a corner, but we have until at least 1.0 to undo that. Waiting to generalize is of course eminently reasonable. My hunch is that we will want to generalize because having `DatasetReaders` fit the `Transform` API will allow us to move reading into the `DataLoader` which gives us much better multiprocess support than our homegrown `MultiprocessDatasetReader`.

Reply:
If I missed something about why this is necessary for integrating with `DatasetReaders`, then by all means, tell me I'm wrong, that's fine.  But, to be more clear, here is a sketch of a proposal that I think we can all agree on:

```python
class Transform(Registrable, Generic[Input, Output]):
    def transform(self, dataset: Iterable[Input]) -> Union[Iterable[Output], Iterable[Iterable[Output]]]:
        raise NotImplementedError
    def transform_batches(self, batches: Iterable[Iterable[Input]]) -> Iterable[Iterable[Output]]:
        raise NotImplementedError
    def __call__(self, dataset: Union[Iterable[Input], Iterable[Iterable[Input]]]) -> ...:
        # base implementation of __call__ goes here

class InstanceTransform(Transform[Instance, Instance]):
    def transform_batches(self, ...):
        # handle case for instance -> instance transforms here

class InstanceToBatchTransform(Transform[Instance, Instance]):
    def transform_batches(self, ...):
        # handle case for instance -> batch transforms here
```

And all type annotations (for `Registrable` / `FromParams` purposes) use `Transform` directly, not either of the bottom two classes.

We're not going to get complete type safety when composing lists of transforms, especially when we read those transforms from a config file.  But that's ok; that's a benefit of using python instead of scala, where this wouldn't actually compile.  And this will be pretty close.

Reply:
So, I tried for a while to get the types to actually work out with what I suggested above, and I couldn't figure it out.  So, here's a better proposal with types that actually work:

```python
class Transform(Registrable, Generic[Input, Output]):
    def transform(self, dataset: Iterable[Input]) -> Iterable[Output]:
        raise NotImplementedError

    def __call__(self, dataset: Union[Iterable[Input], Iterable[Iterable[Input]]) -> Union[...]:
        # This is where we'll hide our type checks to handle the union of possible inputs
        raise NotImplementedError

    # no transform_batches; they types don't really work out to do this right, and it's not needed

class InstanceTransform(Transform[Instance, Instance]):
    # transform() is still abstract, but has type Iterable[Instance] -> Iterable[Instance]

    def __call__(self, Union[...]) -> Union[...]:
        # handle case for instance -> instance transforms here

class InstanceToBatchTransform(Transform[Instance, Iterable[Instance]]):
    # transform() is still abstract, but has type Iterable[Instance] -> Iterable[Iterable[Instance]]

    def __call__(self, ...):
        # handle case for instance -> batch transforms here

class BatchToBatchTransform(Transform[Instance, Iterable[Instance]]):
    # transform() is still abstract, but has type Iterable[Iterable[Instance]] -> Iterable[Iterable[Instance]]

    def __call__(self, ...):
        # handle case for batch -> batch transforms, including throwing an error if not already batched

class FileToInstancesTransform(Transform[str, Instance]):
    # transform() is abstract, but is basically a dataset reader, or calls one

    def __call__(self, dataset: Iterable[str]) -> Iterable[Instance]:
        yield from self.transform(dataset)
```

Reply:
@matt-gardner, thanks a lot for sketching that out! That looks quite good. @DeNeutoy, I think we talked about something like this, but you were concerned that it would cause some duplication because some transforms need to operate on instances directly as well as instances that were already batched. If I understand Matt's code correctly, we would still have some special dispatch in `InstanceTransform.__call__` to handle that case. Alternatively, I think we could have `Transform.__call__` take just `Iterable[Input]` (no `Union`) and provide an explicit `UnpackTransform`. Something like:

```
class UnpackTransform(Transform[Iterable[Instance], Iterable[Instance]]):
    def __init__(self, base_transform: InstanceTransform):
        self._base_transform = base_transform

    def __call__(self, input):
        for batch in input:
            yield self._base_transform(batch)

# In user code
pipeline = [CreateBatches(), ShuffleBatches(), UnpackTransform(SomeInstanceTransform())]
...
```

I slightly prefer being explicit in that manner, but the automatic dispatch seems okay too.

Reply:
Slight correction:

```
class UnpackTransform(BatchTransform):
    def __init__(self, base_transform: InstanceTransform):
        self._base_transform = base_transform

    def transform(self, input: Iterable[Iterable[Instance]]) -> Iterable[Iterable[Instance]]:
        for batch in input:
            yield self._base_transform(batch)
```

Reply:
hmm, @matt-gardner how does this handle the following case:

```
pipeline = [Index(), MaxInstancesInMemory(), Batch()]
pipeline = [Index(), Batch()]
```
Is the idea that there would be two types of batch transforms, one inheriting from `BatchTransform` and one inheriting from `InstanceToBatchTransform`? This seems not great as it means that in a config, you have to know which one to use, which is something I had thought we were explicitly trying to avoid? I'm not sure I see how having the type dispatching in the base `Transform` class would actually avoid this, because if there was only a single class, `Batch[Instance, Iterable[Instance]]`, and it was called with `Iterable[Instance]`, we still don't have strict typing anyway. I think the part i'm missing is how the dispatching would happen without having to have multiple classes?

Reply:
@brendan-ai2 i'm not a massive fan of `UnpackTransform` - it seems better to just have duplicate classes in this case, because I don't think it would be clear to a user which transforms you can unpack. Also, I don't remember having that objection in a conversation - my original conception of this was to have two classes, `Transform` and `BatchTransform`, but Matt convinced me that it was possible to make them order agnostic with this dispatching stuff.

Reply:
Thanks for the correction regarding the motivation. `UnpackTransform` would work on any transform that is an `InstanceTransform` which is clear from the docs/class declaration. If a user is building their own transform pipeline, they'll presumably need to look at the docs a bit regardless. But like I said, automatic dispatch sounds fine to me too.

Reply:
`CreateBatches` is an `InstanceToBatchTransform`, and that class has logic inside of `__call__` to handle the unwrapping that needs to be done when it's given batches as input instead of `Iterable[Instance]`, exactly as you have logic for it in `transform_batch()` now.  We only need one class for this, just as you only need one class in what you have.

There are really just a few simple differences between what you have and what I'm suggesting:
- I'm taking the separate base implementations of `transform_batch` (where you have one of them implemented by default, and the other two overridden manually), and putting them into the `__call__` method of separate base classes.  The concrete transforms all inherit from the appropriate base class, and you basically never have to override `__call__` (except maybe for `StopAfter`, which probably doesn't even need to implement `transform()`).
- I've made the types on the base class explicit with generics, and recommend just annotating uses of these transforms as `List[Transform]` instead of `List[InstanceTransform]`.

I agree with @DeNeutoy on not being a fan of `UnpackTransform` - we should just design the transforms to do the right thing, or crash with a meaningful error if the order you gave doesn't work.

I also would make one small tweak to my proposed API: `__call__` on `Transform` should just have type `Iterable[Input] -> Iterable[Output]`.  We'll break this with `Union` types on some of the subclasses, but that's ok, I think.  The point of separating out `__call__` instead of having everything be in `transform()` is to have a place where we can play loose with the types where it's helpful, while still having `transform()` make sense and be easy to reason about.